File: parseB/lex.cs

using System;
using System.Linq;
using System.Collections.Generic;

namespace ex.parseB
{
    public class Lexer
    {
        private int _index;
        private string _text;

        private readonly char[] _binOpChar = { '.'
                                             , '?'
                                             , '!'
                                             , '+'
                                             , '-'
                                             , '*'
                                             , '/'
                                             , '&'
                                             , '|'
                                             , '^'
                                             , '#'
                                             , '<'
                                             , '>'
                                             , '$'
                                             , '%'
                                             , '~'
                                             , '@'
                                             };


        public IEnumerable<Token> Lex(string input)
        {
            _index = 0;
            _text = input;

            while( !EndText )
            {
                if ( Try(char.IsWhiteSpace) ) 
                {
                    // Nop
                }
                else if ( Try( "//" ) )
                {
                    ClearLine();
                }
                else if ( Try( "/*" ) )
                {
                    ClearBlockComment();
                }
                else if ( Try( '(' ) )
                {
                    yield return new Token { TokenType = TT.LParen };
                }
                else if ( Try( ')' ) )
                {
                    yield return new Token { TokenType = TT.RParen };
                }
                else if ( Try( '[' ) )
                {
                    yield return new Token { TokenType = TT.LSquare };
                }
                else if ( Try( ']' ) )
                {
                    yield return new Token { TokenType = TT.RSquare };
                }
                else if ( Try( '{' ) )
                {
                    yield return new Token { TokenType = TT.LCurly };
                }
                else if ( Try( '}' ) )
                {
                    yield return new Token { TokenType = TT.RCurly };
                }
                else if ( Try( "=" ) )
                {
                    yield return new Token { TokenType = TT.Equal };
                }
                else if ( Try( ':' ) )
                {
                    yield return new Token { TokenType = TT.Colon };
                }
                else if ( Try( ';' ) )
                {
                    yield return new Token { TokenType = TT.SemiColon };
                }
                else if ( Try( ',' ) )
                {
                    yield return new Token { TokenType = TT.Comma };
                }
                else if ( TrySymbol( out var symbol ) )
                {
                    yield return new Token { TokenType = TT.Symbol, SymbolValue = symbol };
                }
                else if ( TryInt( out var value ) )
                {
                    yield return new Token { TokenType = TT.Int, IntValue = value };
                }
                else if ( TryBinOp( out string binOp ) )
                {
                    yield return new Token { TokenType = TT.BinOp, BinOpValue = binOp };
                }
                else
                {
                    throw new Exception( $"Unknown Symbol {Current}" );
                }
            }
        }

        private bool TryBinOp(out string binOp )
        {
            bool IsBinOpChar( char character ) => _binOpChar.Any( c => character == c );

            var bs = new List<char>();
            if ( IsBinOpChar( Current ) )
            {
                bs.Add( Current );
                _index++;
            }
            else
            {
                binOp = "";
                return false;
            }
            while( !EndText && IsBinOpChar( Current ) )
            {
                bs.Add( Current );
                _index++; 
            }
            binOp = new string( bs.ToArray() );
            return true;
        }
    
        private bool TrySymbol(out string symbol)
        {
            var s = new List<char>();
            if ( IsStartSymbolChar( Current ) )
            {
                s.Add( Current );
                _index++;
            }
            else
            {
                symbol = "";
                return false;
            }
            while (!EndText && IsSymbolChar( Current ) )
            {
                s.Add( Current );
                _index++;
            }

            symbol = new string( s.ToArray() );
            return true;
        }

        private bool TryInt( out int value )
        {
            var d = new List<char>();
            if ( char.IsDigit( Current ) )
            {
                d.Add( Current );
                _index++;
            }
            else
            {
                value = 0;
                return false;
            }
            while( !EndText && char.IsDigit( Current ) )
            {
                d.Add( Current );
                _index++; 
            }
            value = int.Parse( new string( d.ToArray() ) );
            return true;
        }
        
        private void ClearLine()
        {
            while( !EndText && Current != '\n' && Current != '\r' )
            {
                _index++;
            }
            if ( !EndText && (Current == '\n' || Current == '\r') )
            {
                _index++;
            }
        }

        private void ClearBlockComment()
        {
            bool EndComment() => Current == '*' && HasNext && Next == '/';
            while( !EndText && !EndComment() )
            {
                _index++;
            }
            if ( !EndText && EndComment() )
            {
                _index+=2;
            }
        }

        private char Previous => _text[_index - 1];
        private char Next => _text[_index + 1];
        private char Current => _text[_index];
        private bool HasPrevious => _index - 1 >= 0;
        private bool HasNext => _index + 1 < _text.Length;
        private bool EndText => _text.Length <= _index; 
        
        private static bool NotSymbolChar( char c )
            => !IsSymbolChar( c );

        private static bool IsStartSymbolChar( char c ) 
            => char.IsLetter( c ) || c == '_';

        private static bool IsSymbolChar( char c ) 
            => char.IsLetterOrDigit( c ) || c == '_';

        private bool Try( char c )
        {
            if ( Current == c )
            {
                _index++;
                return true;
            }
            return false;
        }

        private bool Try( string s )
        {
            var t = _text.Substring( _index );
            if ( t.StartsWith( s ) )
            {
                _index += s.Length;
                return true;
            }
            return false;
        }

        private bool TryKeyword( string s )
        {
            var sub = _text.Substring( _index );

            bool EndOfFile() => sub.Length == s.Length;

            if ( sub.StartsWith( s ) && ( EndOfFile() || NotSymbolChar( sub[s.Length] ) ) )
            {
                _index+=s.Length;
                return true;
            }
            return false;
        }

        private bool Try( Func<char, bool> p ) 
        {
            if ( p( Current ) )
            {
                _index++;
                return true;
            }
            return false;
        }

        private void Is( char c )
        {
            if ( Current != c )
            {
                throw new Exception ( "failure" );
            }
        }
    }
}

File: parseC/Lex.fs
namespace parseC

module Lex = 

    open Data

    let mutable _text = ""
    let mutable _index = 0
    let mutable _buffer = ""


    let _binOpChar = [| '.'
                      ; '?'
                      ; '!'
                      ; '+'
                      ; '-'
                      ; '*'
                      ; '/'
                      ; '&'
                      ; '|'
                      ; '^'
                      ; '#'
                      ; '<'
                      ; '>'
                      ; '$'
                      ; '%'
                      ; '~'
                      ; '@'
                      ; '='
                      ; ':'
                      |]

    let previous = fun () -> _text.[_index - 1]
    let next = fun () -> _text.[_index + 1]
    let current = fun () -> _text.[_index]
    let hasPrevious = fun () -> _index - 1 >= 0;
    let hasNext = fun () -> _index + 1 < _text.Length;
    let endText = fun () -> _text.Length <= _index; 

    let tryChar (c : char) = 
        if c = current () then
            _index <- _index + 1
            true
        else
            false

    let tryWhiteSpace = 
        fun () -> tryChar ' '
               || tryChar '\n'
               || tryChar '\r'
               || tryChar '\t'

    let tryComment = fun () -> 
        if current () = '/' && next () = '/' then
            _index <- _index + 2
            true
        else
            false

    let tryBlockComment = fun () -> 
        if current () = '/' && next () = '*' then
            _index <- _index + 2
            true
        else
            false

    let clearLine = fun () -> 
        while not (endText ()) && current () <> '\n' && current () <> '\r' do
            _index <- _index + 1
        if not (endText ()) && (current () = '\n' || current () = '\r') then 
            _index <- _index + 1

    let clearBlockComment = fun () ->
        let endComment = fun () -> current () = '*' && hasNext () && next () = '/'
        while not (endText ()) && not (endComment()) do 
            _index <- _index + 1
        if not (endText ()) && (endComment ()) then
            _index <- _index + 2
    
    let isDigit (c : char) = System.Char.IsDigit( c )
    let isSymbolChar (c : char) = System.Char.IsLetterOrDigit( c ) || c = '_'
    let isStartSymbolChar (c: char) = System.Char.IsLetter( c ) || c = '_'

    let tryBinOp = fun () ->
        let isBinOpChar (character : char) = Array.exists (fun c -> c = character) _binOpChar

        let mutable bs = "" 
        let add (b:char) = bs <- bs + (string b)

        if isBinOpChar( current () ) then 
            add( current () )
            _index <- _index + 1
            while not (endText ()) && isBinOpChar( current () ) do
                add( current () )
                _index <- _index + 1
            _buffer <- bs
            true
        else
            false

    let trySymbol = fun () ->
        let mutable ls = ""
        let add (l:char) = ls <- ls + (string l)

        if isStartSymbolChar( current () ) then 
            add( current() )
            _index <- _index + 1;
            while not (endText ()) && isSymbolChar( current () ) do
                add( current () )
                _index <- _index + 1
            _buffer <- ls
            true
        else
            false

    let tryInt = fun () ->
        let mutable ds = "" 
        let add (d:char) = ds <- ds + (string d)
        if isDigit( current() ) then 
            add (current ())
            _index <- _index + 1
            while not (endText ()) && isDigit( current () ) do 
                add(current ())
                _index <- _index + 1 
            _buffer <- ds
            true
        else
            false

    let public lex (input : string) =
        _text <- input
        _index <- 0

        let mutable tokens = []
        let add (t:Token) = tokens <- List.Cons( t, tokens )

        while not (endText()) do
            if tryWhiteSpace () then
                ()
            elif tryComment () then
                clearLine ()
            elif tryBlockComment () then
                clearBlockComment ()
            elif tryChar '(' then
                add LParen 
            elif tryChar ')' then
                add RParen 
            elif tryChar '{' then
                add LCurl 
            elif tryChar '}' then
                add RCurl 
            elif tryChar '[' then
                add LSquare
            elif tryChar ']' then
                add RSquare 
            elif tryChar ',' then
                add Comma 
            elif tryChar ';' then
                add SemiColon 
            elif trySymbol () then
                add (Symbol _buffer)
                _buffer <- ""
            elif tryInt () then
                add (Int (int _buffer))
                _buffer <- ""
            elif tryBinOp () then
                add (BinOp _buffer)
                _buffer <- ""
            else  
                raise (Fatal "error")

        List.rev tokens
         